pkg/snmp/gosnmplib/walk.go

```diff
@@ -0,0 +1,127 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package gosnmplib
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/gosnmp/gosnmp"
+)
+
+const (
+	// Base OID for MIB-2 defined SNMP variables
+	// This is the lowest OID that anything should return (most will either be under this prefix or under 1.3.6.1.4.1)
+	baseOID = ".1.3.6.1.2.1"
+	// Java SNMP uses 50, snmp-net uses 10
+	defaultMaxRepetitions = 50
+)
+
+// ConditionalWalk mimics gosnmp.GoSNMP.Walk, except that the walkFn can return
+// a next OID to walk from. Use e.g. SkipOIDRowsNaive to skip over additional
+// rows.
+// This code is adapated directly from gosnmp's walk function.
+func ConditionalWalk(session *gosnmp.GoSNMP, rootOID string, useBulk bool, walkFn func(dataUnit gosnmp.SnmpPDU) (string, error)) error {
+	if rootOID == "" || rootOID == "." {
+		rootOID = baseOID
+	}
+
+	if !strings.HasPrefix(rootOID, ".") {
+		rootOID = string(".") + rootOID
+	}
+
+	oid := rootOID
+	requests := 0
+	maxReps := session.MaxRepetitions
+	if maxReps == 0 {
+		maxReps = defaultMaxRepetitions
+	}
+
+RequestLoop:
+	for {
+		requests++
+		var response *gosnmp.SnmpPacket
+		var err error
+		if useBulk {
+			response, err = session.GetBulk([]string{oid}, 0, maxReps)
```

Comment:
Does it make sense to `useBulk` with a ConditionalWalk? I feel like we would need to get one oid at a time to make sure we compute the right next oid each time.
Can you explain how that would work?

Reply:
It would ignore the returned OIDs for all but the last OID of the bulk; the disadvantage is getting more PDUs than you need; the advantage is that you do fewer requests whenever there are blocks of adjacent OIDs that you do need. I'm not sure when it would be more performant, but e.g. using getbulk for a normal snmpwalk is noticeably much faster in my local instance than using getnext.

The failure case is when you have a lot of consecutive tables that are all fairly small compared to your bulk get size - then you end up fetching most of each table when you only needed to fetch one row from each. But in cases where you have scalars in between the tables you get to make fewer requests, and in cases where you have consecutive tables but they're much larger than your bulk size you don't lose much (e.g. you fetch the first 10 entries of a column from a 200-row table but still get to skip over the remaining 190).

I'm also assuming that the time it takes to process an OID is significantly smaller than the time it takes to send a request to a device and wait for a response - if that's true, then even processing 10x as many OIDs is better if it saves you 10% of your queries.

This could be smarter, though - really I should make it process the first PDU and then skip additional PDUs if they're earlier than the next OID after the first one. I didn't implement that here because comparing OIDs is a bit of a pain, but it shouldn't be too hard to make it do that.